File: services/blockchain-service/src/main/java/com/zerofake/blockchain/service/BlockchainService.java
package com.zerofake.blockchain.service;

import com.zerofake.blockchain.constant.BlockchainStatus;
import com.zerofake.blockchain.constant.TransactionType;
import com.zerofake.blockchain.dto.request.RegisterProductRequest;
import com.zerofake.blockchain.dto.request.TransferOwnershipRequest;
import com.zerofake.blockchain.dto.request.VerifyProductRequest;
import com.zerofake.blockchain.dto.response.BlockchainTransactionResponse;
import com.zerofake.blockchain.dto.response.ProductHistoryResponse;
import com.zerofake.blockchain.dto.response.VerificationResponse;

import java.util.List;
import java.util.UUID;

public interface BlockchainService {

    BlockchainTransactionResponse registerProduct(RegisterProductRequest request);

    BlockchainTransactionResponse transferOwnership(TransferOwnershipRequest request);

    VerificationResponse verifyProduct(VerifyProductRequest request);

    ProductHistoryResponse getProductHistory(UUID productId);

    BlockchainTransactionResponse getTransactionByTransactionId(String transactionId);

    List<BlockchainTransactionResponse> getTransactionsByProductId(UUID productId);

    List<BlockchainTransactionResponse> getTransactionsByStatus(BlockchainStatus status);

    List<BlockchainTransactionResponse> getTransactionsByTransactionType(TransactionType transactionType);

}

File: zerofake-frontend/src/services/blockchainService.ts
import { blockchainApi } from "@/api/blockchainApi";

import type {
  BlockchainTransactionResponse,
  ProductHistoryResponse,
  RegisterProductRequest,
  TransferOwnershipRequest,
} from "@/types/blockchain";

class BlockchainService {
  /**
   * Anchors the product's identity on the ledger.
   *
   * The blockchain service also promotes the product to REGISTERED in the
   * product catalogue, so the client does not need to follow up with a status
   * update of its own.
   */
  async registerProduct(
    request: RegisterProductRequest
  ): Promise<BlockchainTransactionResponse> {
    const response = await blockchainApi.registerProduct(request);
    return response.data;
  }

  async getProductHistory(
    productId: string
  ): Promise<ProductHistoryResponse> {
    const response = await blockchainApi.getProductHistory(productId);
    return response.data;
  }

  async transferOwnership(
    request: TransferOwnershipRequest
  ): Promise<BlockchainTransactionResponse> {
    const response = await blockchainApi.transferOwnership(request);
    return response.data;
  }

  async getTransactionByTransactionId(
    transactionId: string
  ): Promise<BlockchainTransactionResponse> {
    const response =
      await blockchainApi.getTransactionByTransactionId(transactionId);
    return response.data;
  }

  async getTransactionsByProductId(
    productId: string
  ): Promise<BlockchainTransactionResponse[]> {
    const response =
      await blockchainApi.getTransactionsByProductId(productId);
    return response.data ?? [];
  }

  async getAllTransactions(): Promise<BlockchainTransactionResponse[]> {
    const response = await blockchainApi.getAllTransactions();
    return response.data ?? [];
  }
}

export const blockchainService = new BlockchainService();
